### ISSUE TITLE
Attempting to start a Tutorial on Windows desktop with R 4.0.3 results in a console error and an indefinite loading wheel in Tutorial pane.

### ISSUE THREAD

**Issue by R-Hannibal** [rstudio/rstudio#8703]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.3.1093-1 - 1.4.1103-1
    OS Version      : Windows 10
    R Version       : 4.0.3

### Steps to reproduce the problem
1. Uninstall the `nycflighjts13` and `tidyverse` packages from User Library if present.
2. Navigate to the Tutorial pane and attempt to start the Data basics tutorial.
3. Select OK when prompted to install required packages.
4. Select Yes when prompted to restart r prior to installation.

### Describe the problem in detail
An error is displayed in the console: 
```
install.packages(missing)
Error in install.packages : 'match' requires vector arguments
```
A loading wheel appears in the Tutorial pane and continues to spin until the user selects the Home option on the pane.

### Describe the behavior you expected
Required packages are successfully installed and the selected tutorial is launched after install.

### Notes

- This issue does not appear to be a regression as I am able to reproduce on Desktop builds going all the way back to 1.3.1093-1. It appears to be more directly tied to the R version, as when I switch to 3.6.3 on either 1.3 or 1.4 builds I am able to successfully open tutorials.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting a RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by melissa-barca** [comment]
I wasn't able to reproduce this on Windows, but it seems like something might be going on here with R 4.0.3. I had a similar issue testing with Debian where RStudio eventually crashed, and Rich also had a crash with 4.0.3 and RSP on Ubuntu. However, none of these seems to be consistently reproducible.

**Comment by kevinushey** [comment]
I see the following instead, and installation succeeds:

```
> install.packages(c("nycflights13", "tidyverse"))
Installing packages into ‘C:/Users/kevinushey/r/win-library/4.0’
(as ‘lib’ is unspecified)
trying URL 'https://cran.rstudio.com/bin/windows/contrib/4.0/nycflights13_1.0.1.zip'
Content type 'application/zip' length 7110472 bytes (6.8 MB)
downloaded 6.8 MB

trying URL 'https://cran.rstudio.com/bin/windows/contrib/4.0/tidyverse_1.3.0.zip'
Content type 'application/zip' length 439960 bytes (429 KB)
downloaded 429 KB

package ‘nycflights13’ successfully unpacked and MD5 sums checked
package ‘tidyverse’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\tmp\RtmpGg0XJp\downloaded_packages
```

**Comment by kevinushey** [comment]
And here's where the call happens:

https://github.com/rstudio/rstudio/blob/2359fb0ea1318ab27b3af9851c92470ab0d3f31b/src/cpp/session/modules/SessionTutorial.R#L331

@R-Hannibal, can you share the entirety of the output you see when the installation fails?

**Comment by R-Hannibal** [comment]
Sure thing, will include the output from Jobs and console below. Also including a gif of the behavior I'm seeing as it's just a little too odd to describe (notice the odd rapid opening and closing of dialogue windows when it prompts the user to restart R... 🤷‍♂️).

Console output:
```
install.packages(c("nycflights13", "tidyverse"))
Error in install.packages : Updating loaded packages
> install.packages(missing)
Error in install.packages : a character vector argument expected
> install.packages(c("nycflights13", "tidyverse"))
Error in install.packages : Updating loaded packages
> install.packages(missing)
Error in install.packages : a character vector argument expected
> 
```

Jobs output:
```
Error in install_tutorial_dependencies(tutorial_path) : 
  The following packages need to be installed:
  - nycflights13
  - tidyverse
Calls: sourceWithProgress ... eval -> <Anonymous> -> install_tutorial_dependencies
Execution halted
```

I've tested this across both my windows PCs and am getting the same behavior:

![Windows_oddity](https://user-images.githubusercontent.com/62304541/104050078-a423d000-51ab-11eb-9271-4723fb67713a.gif)

**Comment by kevinushey** [comment]
@R-Hannibal: does it make a difference if you click Yes / No rather than closing the dialog when it's displayed?

**Comment by kevinushey** [comment]
I also wonder if you need to restart the IDE after uninstalling those packages (since it looks like other requisite packages were loaded in that session, leading to that dialog being displayed)

**Comment by R-Hannibal** [comment]
@kevinushey - the included gif is with hitting 'No' on the 'Updating Loaded Packages' dialogue window.

Just tried with selecting 'Yes' and it produces the same behavior.

Closing or cancelling the dialogue window produces the following error in the jobs pane and also yields the loading wheel in Tutorial:

```
Error in install_tutorial_dependencies(tutorial_path) :
  The following packages need to be installed:
  - nycflights13
  - tidyverse
Calls: sourceWithProgress ... eval -> <Anonymous> -> install_tutorial_dependencies
Execution halted
```

**Comment by kevinushey** [comment]
Thanks! I think there are a couple issues here:

1. When that dialog is shown, the R code isn't actually halted and so the tutorial tries to run in the background before the requisite packages are installed.

2. The command windows popping up are definitely not expected. My gut feeling is that the newest version of Pandoc has changed something such that, when the tutorial tries to invoke it (to render the tutorial), cmd.exe windows get launched as well.

Both of these are pretty gross, but I don't think they quite meet the bar for 1.4.

**Comment by valerie-rstudio** [comment]
Also experiencing this issue on:
RStudio Edition : Server
RStudio Version : 1.4.1103-1
OS Version : RHEL 7.8
R Version : 4.0.2

**Comment by kevinushey** [comment]
I think the issue here should be alleviated via https://github.com/rstudio/rstudio/commit/74197252b16016bbbf4fc7da6bb18ea03e2c54c3.

The underlying issue is that we replay the code run by the user verbatim in the global environment. You can reproduce something really similar with the following:

```
library(dplyr)

install_deps <- function() {
   pkgs <- "dplyr"
   install.packages(pkgs)
}

install_deps()
```

Assuming `dplyr` is already installed, if you run this code and click "Yes" when prompted to restart, you'll see:

```
> install.packages(pkgs)
Error in install.packages : object 'pkgs' not found
```

... because `pkgs` no longer exists in the environment where we're trying to evaluate `install.packages()`.

**Comment by R-Hannibal** [comment]
Verified Not Fixed on: RStudio Desktop Open Source - 2021.09.0 Build 340, R Version - 4.1.0, Windows 10 - Desktop 

Well, the console error no longer appears and packages are installing successfully, but the tutorial is still failing to load and producing an indefinite loading wheel:

![GO_Tutorial_pane_loading](https://user-images.githubusercontent.com/62304541/131745841-35d845e6-b813-4bda-9366-0483b01ec506.PNG)

After re-launching rstudio and attempting to start the tutorial again, the job fails with the following error text and again the tutorial wheel spins indefinitely `Error: 'getCurrentTheme' is not an exported object from 'namespace:shiny' Execution halted`

![GO_Tutorial_pane_loading_after_relaunch](https://user-images.githubusercontent.com/62304541/131746011-24100237-635a-4d82-8eb0-4370d7c72109.PNG)

**Comment by jmcphers** [comment]
Was the Shiny package auto installed when you started the tutorial, or had it already been installed? 

What do these commands return for you?

```r
> packageVersion("shiny")
> shiny::getCurrentTheme()
```

Does installing the newest version of Shiny (1.6.0 or newer), then restarting RStudio, fix it?

**Comment by R-Hannibal** [comment]
@jmcphers -

Was the Shiny package auto installed when you started the tutorial, or had it already been installed?
        - It had already been installed and was on an earlier version (`1.5` I think).
        - I tried re-testing with removing shiny and letting it be auto-installed when attempting to run the tutorial (this installed 
           shiny version 1.6), still got the hanging/perma-loading behavior on the tutorials pane.

What do these commands return for you?
```
> packageVersion("shiny")
[1] ‘1.6.0’
> shiny::getCurrentTheme()
Error: object ‘faststack’ is not exported by 'namespace:fastmap'
> 
```
This error is the same as what I'm getting in the Jobs pane when attempting to start a tutorial with shiny version 1.6 installed. Tried restarting RStudio, but am consistently getting this error - 

![tutorial_jobs_error](https://user-images.githubusercontent.com/62304541/133120279-a9d6d956-8a5f-49f2-beac-ca62d482aad6.PNG)

**Comment by jmcphers** [comment]
Okay, if that error went away when you installed Shiny 1.6, I think it means we have to bump the Shiny required version to 1.6. 

Now, how about the fastmap package?

    > packageVersion("fastmap")

My hunch is that this is actually a Shiny issue (vaguely it looks like Shiny says it needs fastmap 1.0 but actually requires 1.1).

After recording the version of the fastmap package, could you update it to the latest (1.1) and see if that resolves the issue?

**Comment by R-Hannibal** [comment]
Ah ok, so the version of `fastmap` I had installed was `1.0.1`. When I removed and reinstalled to version `1.1.0`, tutorials now load/run successfully.

One additional oddity, when I attempt to run tutorial without `fastmap` installed, I get an error (included below) in the Jobs pane stating that there is no package called `fastmap` and the tutorial window spins infinitely. Wouldn't the expectation be that the user is offered a prompt to install `fastmap` since it is a dependency?

```
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called 'fastmap'
Calls: sourceWithProgress ... loadNamespace -> withRestarts -> withOneRestart -> doWithOneRestart
Execution halted
```

**Comment by jmcphers** [comment]
Fastmap is a dependency of Shiny, so the only way you should be able to get into this state is to install Shiny and then _uninstall_ Fastmap. However we can make the dependency explicit so this doesn't happen, and it looks like we need to do so anyway since Shiny's dependencies don't declare a requirement for 1.1. (That's a bug in the Shiny package which I'll open separately.)

**Comment by jmcphers** [comment]
Hmmm... now very confused.

- Shiny's dependency on `fastmap` 1.1 did not happen until _after_ Shiny 1.6.0. Shiny 1.6.0 was released in January, and Shiny didn't depend on `fastmap` 1.1 until this commit in February. https://github.com/rstudio/shiny/commit/2b92014ea52132572bdaca18a7dcb4a853fb756b
- `getCurrentTheme` doesn't call `fastmap::faststack` in any codepath.
- I can't repro any of this locally.

There is at least one thing that makes sense. It looks like Shiny R Markdown documents (which learnr uses) now have a dependency on Shiny 1.6.0 due to their use of `getCurrentTheme()`

https://github.com/cran/rmarkdown/blob/767b5107e2b2efa75f23c50cc5fe8f512eb6af63/R/shiny.R#L150

This is an extremely recent development, and kind of an unfortunate one. I hate to require bleeding edge versions of Shiny and all its dependencies; it makes life hard for people on older CRAN snapshots. I'll open a PR and we can discuss in triage.

**Comment by R-Hannibal** [comment]
OK, with shiny package at 1.5.0, and rmarkdown at 2.9, an error is being thrown in the console when attempting to run a Shiny Web Application:

```
> shiny::runApp('Documents/marshinywebapp')
Loading required package: shiny
Warning in as.shiny.appobj.character(appDir) :
  internal error -3 in R_decompress1
Error in as.shiny.appobj.character(appDir) : 
  lazy-load database 'C:/Users/iredh/Documents/R/win-library/4.1/shiny/R/shiny.rdb' is corrupt
Warning in shiny::runApp("Documents/marshinywebapp") :
  internal error -3 in R_decompress1
Error in shiny::runApp("Documents/marshinywebapp") : 
  lazy-load database 'C:/Users/iredh/Documents/R/win-library/4.1/shiny/R/shiny.rdb' is corrupt
Error in shiny::runApp("Documents/marshinywebapp") : 
  lazy-load database 'C:/Users/iredh/Documents/R/win-library/4.1/shiny/R/shiny.rdb' is corrupt
In addition: Warning message:
In shiny::runApp("Documents/marshinywebapp") :
  internal error -3 in R_decompress1
```

**Comment by R-Hannibal** [comment]
I should have been testing this on a Shiny RMarkdown doc (File -> New -> R Markdown -> Shiny), not a shiny web app. 

Attempting top run a Shiny RMarkdown doc with rmarkdown package at 2.1.1 and shiny package at 1.5.0, produces the following error in the render pane:

```
Error: 'getCurrentTheme' is not an exported object from 'namespace:shiny'
Execution halted
```

**Comment by jmcphers** [comment]
That's super unfortunate but confirms our suspicions that we are going to have to require Shiny 1.6 in Ghost Orchid.

**Comment by R-Hannibal** [comment]
Verified Fixed on: RSW - 2021.09.0 Build 344.pro1, R Version - 4.1.0, Operating System - centos7

This issue has taken a few turns along the way, but I have verified that - 

- Windows Specific: Tutorials can now be successfully run on windows when correct packages are installed.
- Non-windows specific: Shiny package version requirement has been bumped to 1.6, which prevents users from getting into a state where they try and do things within the IDE (run a tutorial, or Shiny Web App, or a Shiny RMarkdown file) and get an error due to them using an outdated version of the Shiny package. Now after the latest fix, doing any of these things with an outdated version of the shiny package prompts the user to update and the 1.6.0 version of the shiny package is installed.
